### DISCUSSION TITLE
Non-standard usage

### DISCUSSION THREAD

**Discussion by crossblade** [fasteddy516/CircuitPython_JoystickXL#11] [Q&A] [answered]
Hi!
I have a question about your lib usage. I do design some sort of manipulator on raspberry pi pico. I need it to be HID device (joystick/gamepad/multi-axis). But to solve this I use some sort of multiplexing - only one ADC is used for dozen of resistive sensors (because pico has only 3 ADC inputs). In serial I just switch on GPIO, read value with ADC0, store it in variable. Then repeat this with other GPIOs/sensors. After I do get all the data, do some primitive math and "move joystick". Is it possible to do this with your lib or it is designed to have axis number according to ADCs?

 Thank you in advance

**Comment by fasteddy516** [comment, marked as answer]
The short answer is **yes**, this is possible using JoystickXL.  A word of warning, though - you mentioned "dozen of resistive sensors", but JoystickXL currently supports a *maximum of 8* axes, so if you need more than 8 analog inputs *on a single HID device* then JoystickXL may not work for you.

As far as implementation goes, JoystickXL natively supports ADC expanders that are already supported in CircuitPython such as the [MCP3008](https://learn.adafruit.com/mcp3008-spi-adc/python-circuitpython).  [See the example in the JoystickXL docs here.](https://circuitpython-joystickxl.readthedocs.io/en/latest/examples.html#external-adc)

You can also do completely custom input processing using `VirtualInput` objects as input sources.  If you can collect analog sensor data and convert it to an integer range of 0-65535 (CircuitPython's standard range for analog inputs), JoystickXL can handle the rest.  Here's an example:

```python
"""JoystickXL example using fully custom axis input processing."""
import random

from joystick_xl.inputs import Axis, VirtualInput
from joystick_xl.joystick import Joystick

joystick = Joystick()

joystick.add_input(
    Axis(source=VirtualInput(value=32768)),  # x-axis
    Axis(source=VirtualInput(value=32768)),  # y-axis
    Axis(source=VirtualInput(value=32768)),  # z-axis
    Axis(source=VirtualInput(value=32768)),  # rx-axis
    Axis(source=VirtualInput(value=32768)),  # ry-axis
    Axis(source=VirtualInput(value=32768)),  # rz-axis
    Axis(source=VirtualInput(value=32768)),  # s0-axis
    Axis(source=VirtualInput(value=32768)),  # s1-axis
)


def fancy_multiplexed_input_processing_for_axis(axis: int) -> int:
    """Retrieve and return your multiplexed source values here."""
    new_value = (axis * 8191) + random.randint(0, 8191)  # just generating random data in this example
    return new_value


while True:
    for a, axis in enumerate(joystick.axis):
        axis.source_value = fancy_multiplexed_input_processing_for_axis(a)
    joystick.update()
```

**Reply by crossblade** [reply]
Wow
That works like a charm! Thank you very much for your help
